apps/wing/src/commands/pack.ts

```diff
@@ -0,0 +1,159 @@
+  // check that the library compiles to the "sim" target
+  // TODO: let the user specify all supported targets in package.json
+  console.log('Compiling to the "sim" target...');
+
+  // ugly hack
+  // await compile(".", { target: Target.SIM });
+  await compile(path.join("..", path.basename(process.cwd())), { target: Target.SIM });
+
+  // check npm is installed
+  await checkNpmVersion();
+
+  const userDir = process.cwd();
+  const outdir = resolve(options.outdir);
+
+  // perform our work in a staging directory to avoid making a mess in the user's current directory
+  return withTempDir(async (workdir) => {
+    const excludePaths = [
+      path.join(userDir, "target"),
+      path.join(userDir, "node_modules"),
+      path.join(userDir, ".git"),
+    ];
+
+    // copy the user's directory to the staging directory
+    await copyDir(userDir, workdir, excludePaths);
+
+    // check package.json exists
+    const pkgJsonPath = path.join(workdir, "package.json");
+    if (!(await exists(pkgJsonPath))) {
+      throw new Error(`No package.json found in the current directory. Run \`npm init\` first.`);
+    }
+
+    const pkgJson = JSON.parse(await fs.readFile(pkgJsonPath, "utf8"));
+
+    // check package.json has required fields
+    const requiredFields = ["name", "version", "description", "author", "license"];
+    for (const field of requiredFields) {
+      if (pkgJson[field] === undefined) {
+        throw new Error(`Missing required field "${field}" in package.json.`);
+      }
+    }
+
+    // check that Wing source files will be included in the tarball
+    const pkgJsonFiles = new Set(pkgJson.files ?? []);
+    const expectedGlobs = ["**/*.js", "**/*.w"];
+    for (const glob of expectedGlobs) {
+      if (!pkgJsonFiles.has(glob)) {
+        pkgJsonFiles.add(glob);
+      }
+    }
+    pkgJson.files = [...pkgJsonFiles];
+
+    // check if "main" points to a valid file, and if not, create a dummy file
+    let main = pkgJson.main ?? "index.js";
+    if (!(await exists(main))) {
+      const lines = [];
+      lines.push("// This file was generated by `wing pack`");
+      lines.push(
+        `console.log("${pkgJson.name} is a winglang library and cannot be used from javascript (yet). See winglang.io/docs/libraries")`
+      );
+      lines.push();
+      await fs.writeFile(main, lines.join("\n"));
+    }
+    pkgJson.main = main;
+
+    // add "winglang" to "keywords"
+    const keywords = new Set(pkgJson.keywords ?? []);
+    keywords.add("winglang");
+    pkgJson.keywords = [...keywords];
+
+    // add "wing" to "engines"
+    pkgJson.engines = { wing: "*" };
+
+    // add "wing" top-level field
+    pkgJson.wing = true;
+
+    // write package.json
+    await fs.writeFile(pkgJsonPath, JSON.stringify(pkgJson, null, 2) + "\n");
+
+    // make tarball in the specified outdir
+    await fs.mkdir(outdir, { recursive: true });
+    const command = `npm pack --json --pack-destination "${outdir}"`;
```

Comment:
Any consideration towards doing this programmatically? https://github.com/npm/npm-packlist is what npm and pnpm (and probably yarn but idk) use for packing.

It keeps us compliant with expected npm tarball behavior while not relying on the user's system configuration.

Reply:
I'll try it out

Reply:
Added!

Reply:
Nice!